sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -406,6 +410,250 @@ protected EventProcessor()
+                await using var connectionAwaiter = connection.ConfigureAwait(false);
+
+                var retryPolicy = Options.RetryOptions.ToRetryPolicy();
+                var retryDelay = default(TimeSpan?);
+                var capturedException = default(Exception);
+                var eventBatch = default(IEnumerable<EventData>);
+                var failedAttemptCount = 0;
+                var failedConsumerCount = 0;
+
+                // Continue processing the partition until cancellation is signaled or until the count of failed consumers is too great.
+                // Consumers which been consistently unable to receive and process events will be considered invalid and abandoned for a new consumer.
+
+                while ((!cancellationSource.IsCancellationRequested) && (failedConsumerCount <= MaximumFailedConsumerCount))
+                {
+                    try
+                    {
+                        consumer = CreateConsumer(ConsumerGroup, partition.PartitionId, startingPosition, connection, Options);
+
+                        // Allow the core dispatching loop to apply an additional set of retries over any provided by the consumer
+                        // itself, as a processor should be as resilient as possible and retain partition ownership if processing is
+                        // able to make forward progress.  If the retries are exhausted or a non-retriable exception occurs, the
+                        // consumer will be considered invalid and potentially refreshed.
+
+                        while (!cancellationSource.IsCancellationRequested)
+                        {
+                            try
+                            {
+                                eventBatch = await consumer.ReceiveAsync(EventBatchMaximumCount, Options.MaximumWaitTime, cancellationSource.Token).ConfigureAwait(false);
+                                await ProcessEventBatchAsync(partition, eventBatch, Options.MaximumWaitTime.HasValue, cancellationSource.Token).ConfigureAwait(false);
+
+                                // If event batches are successfully processed, then the need for forward progress is
+                                // satisfied, and the failure counts should reset.
+
+                                failedAttemptCount = 0;
+                                failedConsumerCount = 0;
+                            }
+                            catch (TaskCanceledException) when (cancellationSource.IsCancellationRequested)
+                            {
+                                // Do not log; this is an expected scenario when partition processing is asked to stop.
+
+                                throw;
+                            }
+                            catch (Exception ex) when (!(ex is DeveloperCodeException))
+                            {
+                                _ = OnProcessingErrorAsync(ex, partition, Resources.OperationReadEvents, CancellationToken.None);
```

Comment:
Correct; we're intentionally doing fire-and-forget for the error handlers, as we explicitly state we do not take responsibility for errors in developer-provided code and we do not need to limit to a single concurrent call for the error handler.

We also don't want to signal the error handler to cancel because the processing is being cancelled; the error took place and should be dealt with, which is why I'm using `CancellationToken.None` here.

Reply:
>  If not, could you add a comment as to why we don't await it?

Sure, I'll add a comment.

Reply:
Are there going to be other cases in the Processor where we either await the result of the `Task` returned by `OnProcessingErrorAsync` or cases where we'll pass in a non empty cancellation token?  If not, should we change the signature of this method to just `void OnProcessingError(Exception, TPartition)`?

My concern on returning a task and dropping it on the floor is that if the Task itself faults, it's not clear when or how the failure is observed.

Reply:
The error handler is never observed, by design.  It is always a fire-and-forget call.  This was a topic of much debate during the Event Hubs design, and nobody could come up with something that we felt was a good design; this was the compromise.  I don't think anyone particularly likes it, but it was the best choice that could be agreed upon.

The premise starts with the questions of "if you have a process running in the background, how do you make people aware of errors?"    That was how the error callback came to be.

The follow-up question was "what happens if there is an exception in the callback that handles exceptions?"   You're still running in the background, so you can't directly expose it.  If the error handler is in a bad state where it will always fail, calling it and then catching its error and calling it with that error would lead to a case of either infinite recursion until the process crashes or an infinite loop where you aren't processing events but you're also not actually crashing.

So, the conversation turned to "we will adopt the principal that we do not explicitly handle errors in user code and, instead, let nature take its course."   We wanted to fire-and-forget both handlers, but realized that we had to await the event processing handler because we have to guarantee ordering.  We can't call with a new event for the same partition until the prior one has completed, so that has to be observed, but we don't handle it or route to the error handler, we bubble it up.    

At the end of the day, the architects and service team came to the agreement that it is a reasonable expectation for developers using the processor(s) to accept full responsibility for structing their code to handle errors that could arise from it.   If they do not, we make no guarantees about behavior since we "let nature take its course" and that may differ between runtimes/platforms.

Reply:
Not my preferred design, but the reasoning makes sense.  Thanks, Jesse.